### ISSUE TITLE
Annotate nullability in generated code

### ISSUE THREAD

**Issue by NN---** [protocolbuffers/protobuf#5526]
If it's about generated code change, what programming language?
C#, Java

**Describe the problem you are trying to solve.**
Generated code could be annotated with NotNull , CanBeNull attributes which improves developing experience and save from null value exceptions.

**Describe the solution you'd like**
Generated code with annotations.
They can be either added as external library or just included in the source 
See: 
https://blog.jetbrains.com/dotnet/2015/08/12/how-to-use-jetbrains-annotations-to-improve-resharper-inspections/
https://www.jetbrains.com/help/idea/annotating-source-code.html

**Describe alternatives you've considered**
Alternative is to parse generated code and add annotations by myself.
I don't really like this solution.

**Additional context**

**Comment by BSBandme** [comment]
For java, internally in google protobuf isn't very easy to add new dependency like this (in google we can add guava to mark Nullable annotation if new dependency is allowed). So this improvement may not be seen in the near future.

**Comment by perlun** [comment]
I also agree that this would be useful. We are adding external annotations using `annotations.xml` at the moment but it feels like quite a kludge.

Supporting either  `com.google.code.findbugs:jsr305` (i.e. `javax.annotations`) or Guava annotations would be useful. Perhaps it could be an optional flag? I.e. you could run the protobuf compiler in "traditional" mode, imposing no extra dependency on the generated code whatsoever. This could be the perfect balance between "be conservative with adding new dependencies" and "add functionality generally useful for people".

If we'd go this route, I would even say that being able to configure what non-null annotation class to use would be perfect and a _really_ great approach. (We are using `javax.annotation.Nonnull` ourselves, i.e. JSR305, which plays well with both IntellIJ IDEA and Spotbugs.)

(Note: I have no idea how hard this would be to implement in the Protobuf compiler.)

**Comment by kolmant** [comment]
I was going to post one issue about the same until I found this one. It would be super useful annotating arguments of setters with @NotNull, It throws NPE when null is passed anyway and make working with Protobuf in Kotlin way easier, since Kotlin offers a pretty good null safety when using pure Kotlin or just the Java standard library (and if you're not programming in a defensive way while using the generated code that could be a potential bug since Kotlin does not warn you about the exception).

**Comment by kluever** [comment]
@cpovirk re. nullness annotations

**Comment by cpovirk** [comment]
Only inside Google, though :)

**Comment by googleberg** [comment]
My mistake. I thought these annotations were also in OSS. We should use the standard annotations rather than our custom ones.

**Comment by perlun** [comment]
@googleberg While you're at it, it would be great to get this looked into at some point: https://github.com/protocolbuffers/protobuf/issues/4351#issuecomment-1117284074. It's an incredibly annoying detail of the generated (Java) code where certain methods can in fact return `null`... :grimacing:

**Comment by slovdahl** [comment]
Still very valid issue.

**Comment by git-torrent** [comment]
With edition 2023 and explicit presence, can be (e.g.) C# tweaked to use nullable types generated as a sugar for Has and Clear methods?

**Comment by kluever** [comment]
Keep active.

**Comment by googleberg** [comment]
Still relevant.

**Comment by kluever** [comment]
Keep active.

**Comment by esrauchg** [comment]
Sorry about the slow updates here.

We have concerns with it being understood as a breaking change to even _correctly_ annotate the nullability of our code, and our current stance is that we are intending to avoid doing any breaking change / major version bump on Java Protobuf until 2027.

After internal discussion, our tentative plan to resolve this tension to create our own copies of annotations which exactly match the JSpecify annotations, with the same shortnames but in our a protobuf java package, and begin using that in a minor version bump.

We think this will strike the right balance between unblocking specific nullability-aware tools as long as they either specifically honor our package name or simply look at the shortnames of the annotations, while reducing the risks of it being seen as a breaking change if we used the JSpecify annotations directly. And at a later date, alongside whenever our next Java major version bump we will adopt the "real" JSpecify annotations as the long term state.

If anyone reader has concerns with that being the tentative plan, please let us know. Thanks!

**Comment by NN---** [comment]
Hello,
Does this also apply to .NET annotations ?
They are designed in a way that the breaking changes are minimal.

**Comment by esrauchg** [comment]
Do you mind opening a separate issue specifically about C# so we can relabel this one to be about Java? I feel like when this was previously raised there was maybe a concern about our foundational policy being that our gencode must be compatible with a version of C# that did would fail to compile with them, but I'm failing to find a prior discussion to cite (and even if that was the decision then at some point it would become not-true). The version support [policy](https://github.com/google/oss-policies-info/blob/main/foundational-dotnet-support-matrix.md#new-minor-library-versions) we're subscribed to I think suggests we are still subscribed to 6 and maybe Nullable types is in 8, but we can get a clear answer on a standalone issue.

I'll stress though that the behavior of null objects re: has/clear is a separate topic from simply annotating to document the current intended behavior; we have had a lot of discussions about nullable getters/setters and our current thinking on that is documented here: https://protobuf.dev/design-decisions/nullable-getters-setters/  (which we understand some rational users won't agree with, but we have considered it carefully to arrive at that thinking).
